// @cef-ebsi/ebsi-did-resolver/src/constants.ts
export const EBSI_DID_METHOD_PREFIX = "did:ebsi:";
export const LEGAL_ENTITY = "LEGAL_ENTITY";
export const SUPPORTED_VERSIONS = [LEGAL_ENTITY] as const;
export const EBSI_DID_SPECS: Record<
  (typeof SUPPORTED_VERSIONS)[number],
  { VERSION_ID: number; BYTE_LENGTH: number; PREFIX: string }
> = {
  [LEGAL_ENTITY]: {
    VERSION_ID: 0x01,
    BYTE_LENGTH: 16, // 16 random bytes
    PREFIX: EBSI_DID_METHOD_PREFIX,
  },
} as const;

// The error code from the resolution process.
// https://www.w3.org/TR/did-core/#did-resolution-metadata
export const DID_RESOLUTION_ERROR_CODES = {
  // Values from the specs: https://www.w3.org/TR/did-spec-registries/#error
  INVALID_DID: "invalidDid",
  NOT_FOUND: "notFound",
  REPRESENTATION_NOT_SUPPORTED: "representationNotSupported",
  // Custom values
  UNKNOWN_ERROR: "unknownError",
  INTERNAL_SERVER_ERROR: "internalServerError",
  CONFIGURATION_ERROR: "configurationError",
  INVALID_JWK: "invalidJwk",
} as const;

export type DidResolutionErrorCode =
  (typeof DID_RESOLUTION_ERROR_CODES)[keyof typeof DID_RESOLUTION_ERROR_CODES];
