// @cef-ebsi/did-jwt/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toEthereumAddress = exports.verifyJwt = exports.verifyJWS = exports.ebsiVerifyJwt = exports.decodeJwt = exports.createJwt = exports.NaclSigner = exports.SimpleSigner = void 0;
const SimpleSigner_1 = __importDefault(require("./SimpleSigner"));
exports.SimpleSigner = SimpleSigner_1.default;
const NaclSigner_1 = __importDefault(require("./NaclSigner"));
exports.NaclSigner = NaclSigner_1.default;
const jwt_1 = require("./jwt");
Object.defineProperty(exports, "createJwt", { enumerable: true, get: function () { return jwt_1.createJwt; } });
Object.defineProperty(exports, "decodeJwt", { enumerable: true, get: function () { return jwt_1.decodeJwt; } });
Object.defineProperty(exports, "ebsiVerifyJwt", { enumerable: true, get: function () { return jwt_1.ebsiVerifyJwt; } });
Object.defineProperty(exports, "verifyJWS", { enumerable: true, get: function () { return jwt_1.verifyJWS; } });
Object.defineProperty(exports, "verifyJwt", { enumerable: true, get: function () { return jwt_1.verifyJwt; } });
const crypto_1 = require("./crypto");
Object.defineProperty(exports, "toEthereumAddress", { enumerable: true, get: function () { return crypto_1.toEthereumAddress; } });
