// @cef-ebsi/ebsi-did-resolver/dist/cjs/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProblemDetailsError = void 0;
class ProblemDetailsError extends Error {
    constructor(status, title, options) {
        super(title);
        this.name = "ProblemDetailsError";
        this.status = status;
        this.title = title;
        if (options) {
            if (options.type)
                this.type = options.type;
            if (options.detail)
                this.detail = options.detail;
            if (options.instance)
                this.instance = options.instance;
            if (options.extensions)
                this.extensions = options.extensions;
        }
        if (!this.type)
            this.type = "about:blank";
    }
    toJSON() {
        return Object.assign(Object.assign(Object.assign(Object.assign({ title: this.title, status: this.status }, (this.type && { type: this.type })), (this.detail && { detail: this.detail })), (this.instance && { instance: this.instance })), this.extensions);
    }
    toString() {
        return `${this.status} - ${this.title}`;
    }
}
exports.ProblemDetailsError = ProblemDetailsError;
