// @cef-ebsi/app-jwt/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GRANT_TYPE = exports.Scope = exports.utils = exports.TrustedAppRegistry = exports.Agent = exports.Session = void 0;
const session_1 = __importDefault(require("./session"));
exports.Session = session_1.default;
const agent_1 = __importDefault(require("./agent"));
exports.Agent = agent_1.default;
const trustedAppRegistry_1 = __importDefault(require("./trustedAppRegistry"));
exports.TrustedAppRegistry = trustedAppRegistry_1.default;
const utils = __importStar(require("./utils"));
exports.utils = utils;
const constants_1 = require("./constants");
Object.defineProperty(exports, "GRANT_TYPE", { enumerable: true, get: function () { return constants_1.GRANT_TYPE; } });
Object.defineProperty(exports, "Scope", { enumerable: true, get: function () { return constants_1.Scope; } });
exports.default = {
    Session: session_1.default,
    Agent: agent_1.default,
    TrustedAppRegistry: trustedAppRegistry_1.default,
    utils,
    Scope: constants_1.Scope,
    GRANT_TYPE: constants_1.GRANT_TYPE,
};
