// @cef-ebsi/wallet-lib/src/types.ts
/**
 * JSON Web Key Elliptic Curve ([IANA](https://www.iana.org/assignments/jose/jose.xhtml#web-key-elliptic-curve))
 */
export type JWKEllipticCurve =
  | "P-256"
  | "P-384"
  | "P-521"
  | "Ed25519"
  | "Ed448"
  | "X25519"
  | "X448"
  | "secp256k1";

/**
 * JSON Web Key Operations ([IANA](http://iana.org/assignments/jose/jose.xhtml#web-key-operations))
 */
export type JWKOperation =
  | "sign"
  | "verify"
  | "encrypt"
  | "decrypt"
  | "wrapKey"
  | "unwrapKey"
  | "deriveKey"
  | "deriveBits";

/**
 * JSON Web Key Parameters ([IANA](http://iana.org/assignments/jose/jose.xhtml#web-key-parameters))
 * JWK EC Key
 */
export interface JWKECKey {
  kty: "EC";
  use?: "sig" | "enc";
  key_ops?: JWKOperation[];
  alg?: string;
  kid?: string;
  x5u?: string;
  x5c?: string[];
  x5t?: string;
  "x5t#S256"?: string;
  crv: JWKEllipticCurve;
  x: string;
  y: string;
  d?: string;
}

/**
 * EBSI supported key type, curve and formats
 */
export type EBSIJWKType = "EC";
export type EBSIJWKEllipticCurve = "secp256k1";
export type EBSIKeyFormat = "hex" | "jwk" | "pem";

/**
 * EBSI JWK EC Key
 */
export interface EBSIJWKECKey extends JWKECKey {
  crv: EBSIJWKEllipticCurve;
}
