// @cef-ebsi/ebsi-did-resolver/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DID_RESOLUTION_ERROR_CODES = exports.EBSI_DID_SPECS = exports.SUPPORTED_VERSIONS = exports.NATURAL_PERSON = exports.LEGAL_ENTITY = exports.EBSI_DID_METHOD_PREFIX = void 0;
exports.EBSI_DID_METHOD_PREFIX = "did:ebsi:";
exports.LEGAL_ENTITY = "LEGAL_ENTITY";
exports.NATURAL_PERSON = "NATURAL_PERSON";
exports.SUPPORTED_VERSIONS = [exports.LEGAL_ENTITY, exports.NATURAL_PERSON];
exports.EBSI_DID_SPECS = {
    [exports.LEGAL_ENTITY]: {
        VERSION_ID: 0x01,
        BYTE_LENGTH: 16,
        PREFIX: exports.EBSI_DID_METHOD_PREFIX,
    },
    [exports.NATURAL_PERSON]: {
        VERSION_ID: 0x02,
        BYTE_LENGTH: 32,
        PREFIX: exports.EBSI_DID_METHOD_PREFIX,
    },
};
// The error code from the resolution process.
// https://www.w3.org/TR/did-core/#did-resolution-metadata
exports.DID_RESOLUTION_ERROR_CODES = {
    // Values from the specs: https://www.w3.org/TR/did-spec-registries/#error
    INVALID_DID: "invalidDid",
    NOT_FOUND: "notFound",
    REPRESENTATION_NOT_SUPPORTED: "representationNotSupported",
    // Custom values
    UNKNOWN_ERROR: "unknownError",
    INTERNAL_SERVER_ERROR: "internalServerError",
    CONFIGURATION_ERROR: "configurationError",
    INVALID_JWK: "invalidJwk",
};
